test/functional/treesitter/parser_spec.lua

```diff
@@ -599,6 +599,70 @@ int x = INT_MAX;
 
         eq(result, "value")
       end)
+
+      describe("when setting a key on a capture", function()
+        it("it should create the nested table", function()
+          insert([[
+            int x = 3;
+          ]])
+
+          local result = exec_lua([[
+          local query = require("vim.treesitter.query")
+          local value
+
+          query = vim.treesitter.parse_query("c", '((number_literal) @number (#set! @number "key" "value"))')
+          parser = vim.treesitter.get_parser(0, "c")
+
+          for pattern, match, metadata in query:iter_matches(parser:parse()[1]:root(), 0) do
+            for _, nested_tbl in pairs(metadata) do
+              value = nested_tbl.key
+            end
+          end
+
+          return value
+          ]])
+
+          eq(result, "value")
+        end)
+
+        it("it should not overwrite the nested table", function()
+          insert([[
+            int x = 3;
+          ]])
+
+          local result = exec_lua([[
+          local query = require("vim.treesitter.query")
+          local result = {}
+          result.keys = {}
+          result.values = {}
+
+          query = vim.treesitter.parse_query("c", '((number_literal) @number (#set! @number "key" "value") (#set! @number "key2" "value2"))')
+          parser = vim.treesitter.get_parser(0, "c")
+
+          for pattern, match, metadata in query:iter_matches(parser:parse()[1]:root(), 0) do
+            for _, nested_tbl in pairs(metadata) do
+              for k, v in pairs(nested_tbl) do
+                table.insert(result.keys, k)
+                table.insert(result.values, v)
+              end
```

Comment:
Why no directly returning the table here?

Reply:
I updated the tests to return early 👍